### ISSUE TITLE
[2.0] Describe AssetTargetFallback

### ISSUE THREAD

**Issue by guardrex** [dotnet/docs#2816]
Similar to the request to surface `AssetTargetFallback` in https://github.com/NuGet/docs.microsoft.com-nuget/issues/410, which was prompted by a discussion regarding `PackageTargetFallback` [here](https://github.com/dotnet/cli/issues/3439#issuecomment-319702662), there are a few topics over here that might (probably) need a smidgen of work ... a note? a section? a replacement? versioned content? ... something to surface it:

https://docs.microsoft.com/en-us/dotnet/core/tools/csproj#packagetargetfallback
https://docs.microsoft.com/en-us/dotnet/core/tools/project-json-to-csproj#imports

Cross ref: [Documenting .NET Core MSBuild targets and properties/CSPROJ (2642)](https://github.com/dotnet/docs/issues/2642)

**Comment by GuardRex** [comment]
@mairaw, if you need this and let me know how you want it handled in those two topics, I can square it away tonight. Looks like they both need versioned content.

**Comment by dasMulli** [comment]
Note that especially when upgrading to 2.0, users might see this error if they have been using `PackageTargetFallback `:

```
error NU1003: PackageTargetFallback and AssetTargetFallback cannot be used together. Remove PackageTargetFallback(deprecated) references from the project environment.
```
see https://github.com/dotnet/cli/issues/6952.
TL;DR unless `DisableImplicitAssetTargetFallback` is set to `false`, .NET Core and .NET Standard >= 2.0 projects will implicitly have their `AssetTargetFallback` set to include `net461` (see [source](https://github.com/dotnet/sdk/blob/cd25d6bbfc2aac61cc790c511900427ac64fdaa8/src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets#L139-L141)). 

The only docs I found are in a table at the bottom of https://github.com/NuGet/Home/wiki/Enable-.NET-Core-2.0-projects-to-work-with-.NET-Framework-4.6.1-compatible-packages

**Comment by dasMulli** [comment]
Some more context from https://github.com/dotnet/core/issues/711#issuecomment-312051242

by @rrelyea:

>We have since improved that error to:
PackageTargetFallback is deprecated. Replace PackageTargetFallback references with AssetTargetFallback in the project environment.
>
>More details...PackageTargetFallback worked like imports did in Project.json.
The teams have realized that our implementation of that wasn't ideal.
In order to not break old projects/packages -- we created AssetTargetFallback which is implicitly set for netcoreapp/netstandard2.0 projects via targets.
Some project templates, in the past, used PackageTargetFallback.
When upgrading those projects to v2.0, you end up with both PTF and ATF in effect. Thus we error.
Ideally, you just move to ATF from your existing PTF.

**Comment by mairaw** [comment]
Thank you @guardrex and @dasMulli for the info. I'll try to squeeze this one in asap after I'm done with some other high pri work.

**Comment by dasMulli** [comment]
btw I put up a Q&A post on stack overflow back in august to at least have some info dealing with `NU1003` there and it is still showing activity (visits/votes)
https://stackoverflow.com/questions/45569378/upgrading-to-net-core-2-0-packagetargetfallback-and-assettargetfallback-cannot/45569379

**Comment by guardrex** [comment]
You see @mairaw why I nominated @dasMulli for MVP??!!

He's *soooooo good* that he can answer his own SO questions!

... well ... either that or he's a clone! :alien: :grinning:

**Comment by mairaw** [comment]
Thanks guys! Sorry for the delay. 😢

**Comment by BillWagner** [comment]
raising priority as this has bumped several times.

**Comment by dasMulli** [comment]
btw in 15.8+ MSBuild will also honour the `AssetTargetFallback` value for P2P references - https://github.com/Microsoft/msbuild/pull/3205.

**Comment by ViktorHofer** [comment]
Please document `AssetTargetFallback` as soon as time allows.

**Comment by LinkGarden** [comment]
To help with the right content here:

Projects have one or more targetframeworks.

The target framework determines:
- whether a project reference will be compatible with the hosting project.
- whether a package will be compatible with the hosting project.

If targetframework compatibility rules fail to allow a project reference or a package (via package reference or older techniques), then the system will try again using the assettargetfallback(s) for the hosting project in place of its targetframework(s).

An example:
* Project1 has a targetframework of NetCoreApp2.2
* Project2 has a targetframework of NetCoreApp2.1
* Project3 has a targetframework of NetStandard2
* Project1 has project references to Project2, Project3
* Project1 has a package reference to PackageA which directly supports netcoreapp2.2.
* Project1 has a package reference to PackageB which directly supports net461.
* Project1, after evaluation of the project file, plus the imported props and targets files, has a an assettargetfallback property set to net461.

Because Project1 has a TargetFramework of NetCoreApp2.2 and an AssetTargetFallback of Net461, it has the following results when compatibility testing the packagereferences and projectreferences:
* Project1 -> Project2  --- Compatible due to normal TargetFramework compat rules
* Project1 -> Project3 --- Compatible due to normal TargetFramework compat rules (via netstandard)
* Project1 -> PackageA --- Compatible due to normal TargetFramework compat rules
* Project1 -> PackageB --- NOT COMPATIBLE due to normal TargetFramework compat rules - COMPATIBLE due to AssetTargetFallback.

Happy to help grow this more...

**Comment by rrelyea** [comment]
Created a PR to add some details of AssetTargetFallback to NuGet docs:
https://github.com/NuGet/docs.microsoft.com-nuget/pull/1264
